### ISSUE TITLE
\tcbincludepdf from skins library doesn't work with lualatex

### ISSUE THREAD

**Issue by dbitouze** [T-F-S/tcolorbox#101]
# Context
- OS: Linux (Mageia 7)
- TeX distribution: (up to date) TeX Live 2020
- `tcolorbox`: version 4.30

# Issue
`\tcbincludepdf` from `skins` library works nicely with `pdflatex` and `xelatex` but not with `lualatex`.

## MCE
```latex
\documentclass{article}
\usepackage{tcolorbox}
\tcbuselibrary{skins}

\begin{document}
\tcbincludepdf{example-image-a.pdf}
\end{document}
```
## Error message
```
! Package tcolorbox Error: For PDF support run pdflatex, xelatex, or lualatex.

See the tcolorbox package documentation for explanation.
Type  H <return>  for immediate help.
 ...                                              
                                                  
l.6 \tcbincludepdf{example-image-a.pdf}
                                      
? H
Your compiler seems not to support PDF include files.
? s
OK, entering \scrollmode...
! Undefined control sequence.
<argument> \pdfpages 
          
l.6 \tcbincludepdf{example-image-a.pdf}
                                      
)
Runaway argument?
ndefined\ifpgffor@alphabeticsequence \else \ifpgffor@assign@parse \begingroup \
ETC.
! File ended while scanning use of \pgffor@@dotscharcheck.
<inserted text> 
\par
```
## File List
 ```
*File List*
 article.cls    2019/12/20 v1.4l Standard LaTeX document class
  size10.clo    2019/12/20 v1.4l Standard LaTeX file (size option)
tcolorbox.sty    2020/04/28 version 4.30 text color boxes
     pgf.sty    2020/01/08 v3.1.5b (3.1.5b)
  pgfrcs.sty    2020/01/08 v3.1.5b (3.1.5b)
everyshi.sty    2001/05/15 v3.00 EveryShipout Package (MS)
pgfrcs.code.tex
 pgfcore.sty    2020/01/08 v3.1.5b (3.1.5b)
graphicx.sty    2019/11/30 v1.2a Enhanced LaTeX Graphics (DPC,SPQR)
  keyval.sty    2014/10/28 v1.15 key=value parser (DPC)
graphics.sty    2019/11/30 v1.4a Standard LaTeX Graphics (DPC,SPQR)
    trig.sty    2016/01/03 v1.10 sin cos tan (DPC)
graphics.cfg    2016/06/04 v1.11 sample graphics configuration
  pdftex.def    2018/01/08 v1.0l Graphics/color driver for pdftex
  pgfsys.sty    2020/01/08 v3.1.5b (3.1.5b)
pgfsys.code.tex
pgfsyssoftpath.code.tex    2020/01/08 v3.1.5b (3.1.5b)
pgfsysprotocol.code.tex    2020/01/08 v3.1.5b (3.1.5b)
  xcolor.sty    2016/05/11 v2.12 LaTeX color extensions (UK)
   color.cfg    2016/01/02 v1.6 sample color configuration
pgfcore.code.tex
pgfcomp-version-0-65.sty    2020/01/08 v3.1.5b (3.1.5b)
pgfcomp-version-1-18.sty    2020/01/08 v3.1.5b (3.1.5b)
verbatim.sty    2019/11/10 v1.5r LaTeX2e package for verbatim enhancements
 environ.sty    2014/05/04 v0.3 A new way to define environments
trimspaces.sty    2009/09/17 v1.1 Trim spaces around a token list
etoolbox.sty    2019/09/21 v2.5h e-TeX tools for LaTeX (JAW)
    tikz.sty    2020/01/08 v3.1.5b (3.1.5b)
  pgffor.sty    2020/01/08 v3.1.5b (3.1.5b)
 pgfkeys.sty    
pgfkeys.code.tex
 pgfmath.sty    
pgfmath.code.tex
pgffor.code.tex
tikz.code.tex
l3backend-pdfmode.def    2020-05-05 L3 backend support: PDF mode
supp-pdf.mkii
epstopdf-base.sty    2020-01-24 v2.11 Base part for package epstopdf
epstopdf-sys.cfg    2010/07/13 v1.3 Configuration of (r)epstopdf for TeX Live
example-image-a.pdf    Graphic file (type pdf)
 ***********
```

**Comment by muzimuzhi** [comment]
Try this:

```tex
\documentclass{article}
\usepackage{tcolorbox}
\tcbuselibrary{skins}

\makeatletter
\def\tcb@set@pdfpages#1#2{%
  \IfFileExists{#2}{%
    \ifdef{\pdfximage}{%
      % pdftex
      \pdfximage{#2}%
      \edef#1{\the\pdflastximagepages}%
    }{%
      \ifdef{\saveimageresource}{%
        % luatex
        \saveimageresource{#2}%
        \edef#1{\the\lastsavedimageresourcepages}%
      }{%
        \ifdef{\XeTeXpdfpagecount}{%
          % xetex
          \edef#1{\the\XeTeXpdfpagecount #2 }%
          \ifnum#1=0 %
            \edef#1{1}%
          \fi%
        }{\tcb@error{For PDF support run pdflatex, xelatex, or lualatex}{%
          Your compiler seems not to support PDF include files.}}%
      }%
    }%
  }{\tcb@error{PDF file `#2' not found}{The PDF file `#2' was not found}}%
  \relax%
}
\makeatother

\begin{document}
\tcbincludepdf{example-image-a.pdf}
\end{document}
```

The names of primitives used to get the pages of a PDF file vary among engines. 
 - The current implementation counts only pdftex and xetex, but not luatex. 
 - My above example adds the support for luatex, using the name mapping in the following table, on [LuaTeX Reference Manual](http://mirrors.ctan.org/systems/doc/luatex/luatex.pdf) v1.12,  Sec. 3.1.4, page 42.

![image](https://user-images.githubusercontent.com/6376638/82723246-1ff60880-9d00-11ea-975d-1aba8ae94303.png)

**Comment by muzimuzhi** [comment]
With the naming normalization provided by LaTeX3 (for pdftex and luatex), this can be simplified a bit

```tex
\pdftex_pdfximage:D 
  -> \pdfximage and \saveimageresource
\tex_pdflastximagepages:D 
  -> \pdflastximagepages and \lastsavedimageresourcepages
```

```tex
\documentclass{article}
\usepackage{tcolorbox}
\tcbuselibrary{skins}

\makeatletter
\ExplSyntaxOn
\def\tcb@set@pdfpages#1#2{%
  \IfFileExists{#2}{%
    \ifdef{\pdftex_pdfximage:D}{%
      % pdftex or luatex
      \pdftex_pdfximage:D{#2}%
      \edef#1{\the\tex_pdflastximagepages:D}%
    }{%
      \ifdef{\XeTeXpdfpagecount}{%
        % xetex
        \edef#1{\the\XeTeXpdfpagecount #2~}% NOTE the ~ used as a space char
        \ifnum#1=0 %
          \edef#1{1}%
        \fi%
      }{\tcb@error{For PDF support run pdflatex, xelatex, or lualatex}{%
        Your compiler seems not to support PDF include files.}}%
    }%
  }{\tcb@error{PDF file `#2' not found}{The PDF file `#2' was not found}}%
  \relax%
}
\ExplSyntaxOff
\makeatother

\begin{document}
\tcbincludepdf{example-image-a.pdf}
\end{document}
```

**Comment by dbitouze** [comment]
Nice! Ready for a PR? :smile: BTW, how did you proceed to unearth this nice `\pdftex_pdfximage:D`?

**Comment by muzimuzhi** [comment]
> Nice! Ready for a PR? 😄 

The [commit history](https://github.com/T-F-S/tcolorbox/commits/master) implies that this repo is more like a releasing place, than a developing one. So I hesitate to open a PR.

> BTW, how did you proceed to unearth this nice `\pdftex_pdfximage:D`?

Already known that latex3 has aliased every primitive, I searched in `texdoc source3`, the documented source code of latex3. Maybe a table showing the primitive and aliased name(s) would be helpful for latex3 users.

**Comment by joulev** [comment]
> > Nice! Ready for a PR? 😄
> 
> The [commit history](https://github.com/T-F-S/tcolorbox/commits/master) implies that this repo is more like a releasing place, than a developing one. So I hesitate to open a PR.
> 
> > BTW, how did you proceed to unearth this nice `\pdftex_pdfximage:D`?
> 
> Already known that latex3 has aliased every primitive, I searched in `texdoc source3`, the documented source code of latex3. Maybe a table showing the primitive and aliased name(s) would be helpful for latex3 users.

I doubt the LaTeX3 kernel team will ever do that though: they have explicitly stated that functions with the D argument should not be used at the user level (they even say no one but the kernel team should use such functions, but imho that is way too restrictive and I still sometimes use them).

**Comment by T-F-S** [comment]
Thank you all for reporting and solving the issue. I will take @muzimuzhi first suggestion more or less unchanged for the next release, because it avoids the LaTeX3 `D` function (even if this code is more elegant).

**Comment by muzimuzhi** [comment]
Note: resolved in v4.31 by https://github.com/T-F-S/tcolorbox/commit/55d9deed444eef58c65a88b893500255f44ac7ba, more specifically [`these diff lines`](https://github.com/T-F-S/tcolorbox/commit/55d9deed444eef58c65a88b893500255f44ac7ba#diff-fa2af706d885320ba3f6b697e5c1bdd4R1813-R1826).

**Comment by dbitouze** [comment]
@muzimuzhi Works like a charm now, thanks!
